kyo-stm/shared/src/main/scala/kyo/TMap.scala

```diff
@@ -0,0 +1,280 @@
+package kyo
+
+/** A transactional map implementation that provides atomic operations on key-value pairs within STM transactions. Internally represented as
+  * `TRef[Map[K, TRef[V]]]`, where each value is wrapped in its own transactional reference.
+  *
+  * TMap is designed to minimize contention in concurrent scenarios through this nested TRef structure. Since each value has its own `TRef`,
+  * operations on different keys can proceed independently. Only structural changes like `put` with new keys or `remove` need to lock the
+  * main map, while value updates through `updateWith` or `put` to existing keys only lock that specific value's TRef.
+  *
+  * This architecture is particularly effective at reducing retries in concurrent scenarios by limiting the scope of conflicts between
+  * transactions. Updates to different keys can proceed in parallel, while reads can execute concurrently with writes to different keys. The
+  * implementation maintains strong consistency guarantees while allowing maximum concurrency for non-conflicting operations.
+  *
+  * Operations that modify existing values have low contention characteristics, while structural modifications experience higher contention.
+  * This makes TMap particularly well-suited for scenarios with high concurrent access to different keys where most operations are reads or
+  * updates to existing values.
+  */
+opaque type TMap[K, V] = TRef[Map[K, TRef[V]]]
```

Comment:
This approach is different from ZIO. It provides more granular locking with the goal to reduce transaction conflicts and retries. It seems ZIO's `TMap` requires locking the map itself in all modifications: https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/stm/TMap.scala

Reply:
Are you referring to the lock in the zio implementation? It's only used on resizing the array which doesn't happen on all modifications.

Reply:
Based on my reading of ZIO's code, any modifications to `TMap` require a write to the underlying `TArray` reference, which implies a lock at commit time of the map and essentially serializes modifications during commit. In this new encoding of `TMap`, each map entry has its own `TRef`. 

For example, updating an existing key requires only a read lock of the outer map `TRef` (to validate the structure) and a write lock of just that entry's specific `TRef`. This means that concurrent modifications to different existing keys can commit in parallel while reads of other keys can also commit in parallel, with only structural changes forcing full commit-time serialization. The tradeoff is memory overhead since each value needs its own `TRef` but this initial `TRef` implementation is already quite lightweight. 

I'm planning to work on benchmarks to validate these tradeoffs, particularly around concurrent access patterns.